UIforETW/Utility.cpp

```diff
@@ -594,69 +871,171 @@ void SetCurrentThreadName(PCSTR threadName)
+	copyWPAProfileToDocuments(force);
+
+	copyWPAProfileToLocalAppData(exeDir, force);
+}
+
+_Success_(return)
+bool OpenRegKey( _Out_ HKEY* const key, _In_ const HKEY root, PCWSTR const subkey )
+{
+	const LONG openResult = ::RegOpenKeyExW(root, subkey, 0, KEY_ALL_ACCESS, key);
+	if (openResult == ERROR_SUCCESS)
+		return true;
+	outputPrintf(L"Failed to open registry key `%s`.\n", subkey);
+	debugLastError();
+	return false;
+}
+
+void CloseFindHandle(_In_ _Pre_valid_ _Post_ptr_invalid_ const HANDLE handle, PCWSTR const directory)
+{
+	const BOOL findClose = ::FindClose(handle);
+	if (findClose)
+		return;
+	outputPrintf(L"FindClose (for directory: `%s`) failed.\n", directory);
+	debugLastError();
+	std::terminate();
+}
+
+void CloseValidHandle(_In_ _Pre_valid_ _Post_ptr_invalid_ const HANDLE handle)
+{
+	const BOOL handleClosed = ::CloseHandle(handle);
+	if (handleClosed == 0)
+		std::terminate();//Logic bug!
+}
+
+void CloseRegKey(_In_ _Pre_valid_ _Post_ptr_invalid_ const HKEY key, PCWSTR const keyName)
+{
+	const LONG closeKey = ::RegCloseKey(key);
+	if (closeKey == ERROR_SUCCESS)
+		return;
+	outputPrintf(L"Failed to close registry key `%s`.\n", keyName);
+	debugLastError();
+}
+
+void ClipboardClose()
+{
+	const BOOL closeResult = ::CloseClipboard( );
+	if (closeResult)
+		return;
+	debugPrintf(L"Failed to close the clipboard!\n");
+	debugLastError( );
+}
+
+bool ClipboardOpen()
+{
+	const BOOL openClip = ::OpenClipboard(::GetDesktopWindow());
+	if (!openClip)
 	{
-		std::wstring source = exeDir + fileName;
-		std::wstring destDir = documents + std::wstring(L"\\WPA Files");
-		std::wstring dest = destDir + fileName;
-		if (force || !PathFileExists(dest.c_str()))
-		{
-			(void)_wmkdir(destDir.c_str());
-			if (CopyFile(source.c_str(), dest.c_str(), FALSE))
-			{
-				if (force)
-					outputPrintf(L"Copied Startup.wpaProfile to the WPA Files directory.\n");
-			}
-			else
-			{
-				if (force)
-					outputPrintf(L"Failed to copy Startup.wpaProfile to the WPA Files directory.\n");
-			}
-		}
+		outputPrintf(L"Failed to open clipboard!\n");
+		debugLastError();
+		return false;
 	}
+	return true;
+}
 
-	// Then copy the WPA 10 startup.wpaProfile file
-#pragma warning(suppress : 4996)
-	const wchar_t* localAppData = _wgetenv(L"localappdata");
-	if (localAppData)
-	{
-		std::wstring source = exeDir + L"\\startup10.wpaProfile";
-		std::wstring destDir = std::wstring(localAppData) + L"\\Windows Performance Analyzer";
-		std::wstring dest = destDir + fileName;
-		if (force || !PathFileExists(dest.c_str()))
-		{
-			(void)_wmkdir(destDir.c_str());
-			if (CopyFile(source.c_str(), dest.c_str(), FALSE))
-			{
-				if (force)
-					outputPrintf(L"%s", L"Copied Startup.10wpaProfile to %localappdata%\\Windows Performance Analyzer\n");
-			}
-			else
-			{
-				if (force)
-					outputPrintf(L"%s", L"Failed to copy Startup.10wpaProfile to %localappdata%\\Windows Performance Analyzer\n");
-			}
-		}
+bool ClipboardEmpty()

```

Comment:
This is a badly named function. "ClipboardEmpty()" sounds like it queries the clipboard to see if it is empty. It does not sound like it empties the clipboard.